pkg/cri/server/sandbox_run.go

```diff
@@ -218,16 +192,34 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
 		containerd.WithContainerExtension(sandboxMetadataExtension, &sandbox.Metadata),
 		containerd.WithRuntime(ociRuntime.Type, runtimeOpts)}
 
+	// A flag to keep the container in metadata store even in error case.
+	// Combined with adding sandbox to sandboxstore, it allows kubelet to reconcile the sandbox status for cleanup.
+	// e.g. when teardownPodNetwork fails
+	shouldDelContainer := true
+
 	container, err := c.client.NewContainer(ctx, id, opts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create containerd container: %w", err)
 	}
 	defer func() {
 		if retErr != nil {
-			deferCtx, deferCancel := ctrdutil.DeferContext()
-			defer deferCancel()
-			if err := container.Delete(deferCtx, containerd.WithSnapshotCleanup); err != nil {
-				log.G(ctx).WithError(err).Errorf("Failed to delete containerd container %q", id)
+			if shouldDelContainer {
+				deferCtx, deferCancel := ctrdutil.DeferContext()
+				defer deferCancel()
+				if err := container.Delete(deferCtx, containerd.WithSnapshotCleanup); err != nil {
```

Comment:
validate container here before calling Delete..

Reply:
Sorry, not clear what validation you refer to. Can you elaborate?